drivers/flash/flash_rpc/flash_rpc_controller.c

```diff
@@ -0,0 +1,240 @@
+NRF_RPC_GROUP_DEFINE(flash_rpc_api, "flash_rpc_api", &flash_rpc_api_tr, NULL, NULL, NULL);
+
+#if DT_NODE_HAS_STATUS(DT_INST(0, nordic_rpc_flash_controller), okay)
+#define DT_DRV_COMPAT nordic_rpc_flash_controller
+#else
+#error "Node is not available"
+#endif
+
+#define FLASH_RPC_CONTROLLER_NODE DT_INST(0, nordic_rpc_flash_controller)
+#define FLASH_RPC_NODE DT_INST_CHILD(0, flash_rpc_0)
+#define FLASH_RPC_FLASH_SIZE DT_REG_SIZE(FLASH_RPC_NODE)
+#define FLASH_RPC_ERASE_UNIT DT_PROP(FLASH_RPC_NODE, erase_block_size)
+#define FLASH_RPC_PROG_UNIT DT_PROP(FLASH_RPC_NODE, write_block_size)
+#define FLASH_RPC_FLASH_SIZE DT_REG_SIZE(FLASH_RPC_NODE)
+
+#define FLASH_RPC_PAGE_COUNT (FLASH_RPC_FLASH_SIZE/FLASH_RPC_ERASE_UNIT)
+
+static const struct flash_parameters flash_rpc_parameters = {
+	.write_block_size = FLASH_RPC_PROG_UNIT,
+	.erase_value = 0xff,
+};
+
+void rsp_error_code_handle(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,
+			   void *handler_data)
+{
+	int32_t val;
+
+	if (zcbor_int32_decode(ctx->zs, &val)) {
+		*(int *)handler_data = (int)val;
+	} else {
+		*(int *)handler_data = -NRF_EINVAL;
+	}
+}
+
+static void flash_rpc_get_rsp(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,
+				 void *handler_data)
+{
+	int *result = (int *)handler_data;
+
+	if (!zcbor_int32_decode(ctx->zs, result)) {
+		LOG_ERR("Unable to decode result");
+		*result = -EINVAL;
+	}
+}
+
+static bool encode_flash_msg(struct nrf_rpc_cbor_ctx *ctx, off_t *offset, void *ptr, size_t *len)
+{
+	NRF_RPC_CBOR_ALLOC(&flash_rpc_api, *ctx, CBOR_BUF_FLASH_MSG_SIZE);
+
+	if (!zcbor_uint32_put(ctx->zs, (uint32_t)*offset)) {
+		return false;
+	}
+
+	if (!zcbor_uint32_put(ctx->zs, (uint32_t)ptr)) {
+		return false;
+	}
+
+	return zcbor_uint32_put(ctx->zs, (uint32_t)*len);
+}
+
+#ifndef CONFIG_FLASH_RPC_SYS_INIT
+static void err_handler(const struct nrf_rpc_err_report *report)
+{
+	LOG_ERR("nRF RPC error %d ocurred. See nRF RPC logs for more details.",
+	       report->code);
+	k_oops();
+}
+#endif
+
+int flash_rpc_init(const struct device *dev)
+{
+	int err;
+	int result;
+	struct nrf_rpc_cbor_ctx ctx;
+	const struct rpc_flash_config *dev_config = dev->config;
+
+	ARG_UNUSED(dev_config);
+
+#ifndef CONFIG_FLASH_RPC_SYS_INIT
+	err = nrf_rpc_init(err_handler);
+	if (err) {
+		LOG_ERR("Initializing nRF RPC failed: %d", err);
+		return -EINVAL;
+	}
+#endif
+
+	NRF_RPC_CBOR_ALLOC(&flash_rpc_api, ctx, CBOR_BUF_FLASH_MSG_SIZE);
+
+	err = nrf_rpc_cbor_cmd(&flash_rpc_api, RPC_COMMAND_FLASH_INIT, &ctx,
+			       rsp_error_code_handle, &result);
+	if (err != 0) {
+		LOG_ERR("Could not send RPC command: %d", err);
+		return err;
+	}
+
+	return result;
+}
+
+int flash_rpc_read(const struct device *dev, off_t offset, void *buffer, size_t len)
+{
+	ARG_UNUSED(dev);
+	int err;
+	int result;
+
+	struct nrf_rpc_cbor_ctx ctx;
+
+	if (buffer == NULL || len == 0) {
+		return -EINVAL;
+	}
```

Comment:
This is disputable, but some of our drivers just return 0 and do nothing when `len == 0`. I think that this is good approach as there is no need to error if we are not going to do any work.